java-engine/src/main/java/io/getunleash/engine/UnleashEngine.java

```diff
@@ -1,62 +1,86 @@
 package io.getunleash.engine;
 
 import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.core.type.TypeReference;
 import com.fasterxml.jackson.databind.DeserializationFeature;
 import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.ObjectReader;
+import com.fasterxml.jackson.databind.ObjectWriter;
+import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
 import com.sun.jna.Pointer;
 
+import java.io.IOException;
+
 public class UnleashEngine {
     private static final String UTF_8 = "UTF-8";
     private final YggdrasilFFI yggdrasil;
-    private final ObjectMapper mapper;
+    private final ObjectReader reader;
+    private final ObjectWriter writer;
 
     public UnleashEngine() {
         this(new YggdrasilFFI());
     }
 
     UnleashEngine(YggdrasilFFI yggdrasil) {
         this.yggdrasil = yggdrasil;
-        mapper = new ObjectMapper();
+        ObjectMapper mapper = new ObjectMapper();
+        mapper.registerModule(new JavaTimeModule());
         mapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
+        reader = mapper.reader();
+        writer = mapper.writer();
     }
 
     public void takeState(String toggles) throws YggdrasilInvalidInputException {
-        TakeStateResponse response = read(yggdrasil.takeState(toggles), TakeStateResponse.class);
+        YggResponse<Void> response = read(yggdrasil.takeState(toggles), new TypeReference<>() {});
         if (!response.isValid()) {
             throw new YggdrasilInvalidInputException(toggles);
         }
     }
 
-    public boolean isEnabled(String name, Context context) throws YggdrasilInvalidInputException {
+    public Boolean isEnabled(String name, Context context) throws YggdrasilInvalidInputException, YggdrasilError {
         try {
-            String jsonContext = mapper.writeValueAsString(context);
-            IsEnabledResponse isEnabled = read(yggdrasil.checkEnabled(name, jsonContext), IsEnabledResponse.class);
-            return isEnabled.isEnabled();
+            String jsonContext = writer.writeValueAsString(context);
+            YggResponse<Boolean> isEnabled = read(yggdrasil.checkEnabled(name, jsonContext), new TypeReference<>() {});
+            return isEnabled.getValue();
         } catch (JsonProcessingException e) {
             throw new YggdrasilInvalidInputException(context);
         }
     }
 
-    public VariantResponse getVariant(String name, Context context) throws YggdrasilInvalidInputException {
+    public VariantDef getVariant(String name, Context context) throws YggdrasilInvalidInputException, YggdrasilError {
```

Comment:
I'm not sure if I want to start having too many exceptions in this interface... :thinking: 
- `YggdrasilInvalidInputException`: for bad user input
- `YggdrasilError`: something went terribly wrong

I believe I can just have YggdrasilError and have YggdrasilInvalidInputException extending from it, but I'd like to get your input @daveleek @sighphyre

Reply:
Maybe there's a tiny discussion to be had around whether or not we use the name "Yggdrasil" in the wrapper/platform specific engine - and leaking it out to SDKs. That it should be the same across all wrappers and SDKs?

I know TakeState will throw a bad json error on TakeState, would that be happening here as well with the Context @sighphyre ?

Reply:
WRT takeState and bad input I'm mapping this into a checked exception: YggdrasilInvalidInputException. Not sure we want standardization across Engines at this stage.

About Yggdrassil leaking outside, I like to make it visible, but also understand your reasons, not sure...

Reply:
Yeah take state can go bad in theory but it _shouldn't_ since we control both sides. If we're worried about too many exceptions for folks to have to think about we can always have one base Yggdrasil exception class and have child classes. Gives people enough granularity to catch specific things if they care and anything if they don't

Reply:
Ok, I'll keep it this way for now, as it will only be consumed by our SDK which we control. If we feel the pain, we change it later. Maybe next iteration we can have the hierarchy of exceptions. I think that'd be a good solution